File: src/database/SchemaManager.cpp
#include "database/SchemaManager.hpp"
#include "database/DatabaseException.hpp"

#include <fstream>
#include <sstream>
#include <pqxx/pqxx>

namespace reconciliation::database {

namespace {
std::string readFile(const std::string& path) {
    std::ifstream file(path);
    if (!file.is_open()) {
        throw std::runtime_error("SchemaManager: could not open " + path);
    }
    std::ostringstream oss;
    oss << file.rdbuf();
    return oss.str();
}
} // namespace

void SchemaManager::initializeSchema(DatabaseConnection& connection,
                                      const std::string& schemaPath,
                                      const std::string& indexesPath) {
    try {
        std::string schemaSql = readFile(schemaPath);
        pqxx::work schemaTxn(connection.handle());
        schemaTxn.exec(schemaSql);
        schemaTxn.commit();

        std::string indexesSql = readFile(indexesPath);
        pqxx::work indexTxn(connection.handle());
        indexTxn.exec(indexesSql);
        indexTxn.commit();
    } catch (const pqxx::sql_error& e) {
        throw DatabaseException(std::string("SchemaManager::initializeSchema failed: ") + e.what());
    }
}

} // namespace reconciliation::database

File: recon_engine/database/schema_manager.py
from __future__ import annotations

import os

import psycopg2

from .connection import DatabaseConnection, DatabaseException

_SQL_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "sql")


class SchemaManager:
    @staticmethod
    def initialize_schema(conn: DatabaseConnection) -> None:
        schema_path = os.path.join(_SQL_DIR, "schema.sql")
        indexes_path = os.path.join(_SQL_DIR, "indexes.sql")
        try:
            with open(schema_path) as f:
                schema_sql = f.read()
            with open(indexes_path) as f:
                indexes_sql = f.read()

            with conn.handle.cursor() as cur:
                cur.execute(schema_sql)
                cur.execute(indexes_sql)
            conn.handle.commit()
        except (psycopg2.Error, OSError) as e:
            conn.handle.rollback()
            raise DatabaseException(f"initializeSchema failed: {e}") from e
